### PR TITLE
Harbor addon

### PR DESCRIPTION
**Pull request by vincentni**
*Description of changes:*
the following images are built:
harbor/harbor-core harbor/harbor-db harbor/harbor-exporter harbor/harbor-jobservice harbor/harbor-log harbor/harbor-nginx harbor/harbor-portal harbor/harbor-redis harbor/harbor-registry harbor/harbor-registryctl

the remaining images to be built:
harbor/harbor-chart-server harbor/harbor-notary-signer harbor/harbor-notary-server harbor/harbor-trivy-adapter

### CODE REVIEW

File: Common.mk
Comment (vincentni) [inline Common.mk:50]: This trick is a workaround for project structure not aligning with git repo structure because I put all harbor-related components under the same parent folder (goharbor).

File: Common.mk
Comment (vincentni) [inline Common.mk:321]: Harbor needs to generate some swagger code before compilation otherwise it will fail.

File: projects/goharbor/harbor-notary/GIT_TAG
Comment (vincentni) [inline projects/goharbor/harbor-notary/GIT_TAG:1]: Harbor uses notary v0.6.1 that doesn't have go.mod file, which is required in our build-tooling. So I bump that up but need to be tested.

File: projects/goharbor/harbor/Makefile
Comment (vincentni) [inline projects/goharbor/harbor/Makefile:87]: registry compiles when I put its source code along with harbor main source code. But I can pull that out into a new project and make it as a dependency.

File: projects/goharbor/harbor-chart-server/Makefile
Comment (vincentni) [inline projects/goharbor/harbor-chart-server/Makefile:25]: It is to accommodate project structure not aligning with git repo structure so build-tooling can still pull code from right place and generate output to the right place.

File: projects/goharbor/harbor-chart-server/Makefile
Comment (jaxesn) [inline projects/goharbor/harbor-chart-server/Makefile:1]: Lets break this project out into another PR and start there and try to get this one all the way thru, code + prow jobs + cdk changes + etc.  There looks like some oddities here but for the most part doesn't look to nonstandard so probably a decent one to start with to get through the whole process.

Comment (TerryHowe) [reply projects/goharbor/harbor-chart-server/Makefile:1]: particularly this one since we are handling helm charts differently anyway

File: projects/goharbor/harbor/Makefile
Comment (vincentni) [reply projects/goharbor/harbor/Makefile]: registry has been factored out to a new project distribution/distribution

File: projects/distribution/distribution/Makefile
Comment (jaxesn) [inline projects/distribution/distribution/Makefile:16]:
```suggestion
EXTRA_GOBUILD_FLAGS=-tags include_oss include_gcs
```

Quotes may be necessary around the two tags, which can be a bit of pain to make sure they flow through make correctly.

File: projects/goharbor/harbor/Makefile
Comment (jaxesn) [inline projects/goharbor/harbor/Makefile:46]: Lets make this var all caps for consistency.

File: projects/goharbor/harbor/Makefile
Comment (jaxesn) [inline projects/goharbor/harbor/Makefile:103]: similar comment as above

File: projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile
Comment (jaxesn) [inline projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile:1]: If we are using the standard base, which I would rather we didnt have to since it increases your CVE landscape due to a larger number of OS packages you dont even need, then you do not to do all the useradd stuff in the makefile.  Instead you can do it like you did here using shaddow-utils.  I would probably remove shadow utils in the same layer after creating the user/group just to avoid it being in the final image.

File: projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile
Comment (jaxesn) [inline projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile:12]: Want to move this to the bottom of the file?

File: projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile
Comment (jaxesn) [inline projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile:24]: I think you can avoid this by passing --chown to the COPY call above.

File: projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile
Comment (jaxesn) [inline projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile:25]: Do these scripts not already have the ux flags?

File: projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile
Comment (jaxesn) [inline projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile:15]: We dont usually do this. Could we copy these files into output like youve done for others.

File: projects/goharbor/harbor/docker/linux/harbor-nginx/Dockerfile
Comment (jaxesn) [inline projects/goharbor/harbor/docker/linux/harbor-nginx/Dockerfile:1]: We have an nginx base now you could probably use.

File: projects/goharbor/harbor/docker/linux/harbor-redis/Dockerfile
Comment (jaxesn) [inline projects/goharbor/harbor/docker/linux/harbor-redis/Dockerfile:18]: For this one do we really need the healthcheck script?  Kubernetes isnt going to call this one anyway right?  If we dont need this, can we use the minimal base here and install redis like we did [here](https://github.com/aws/eks-anywhere-build-tooling/blob/main/projects/redis/redis/docker/linux/Dockerfile).

**Review by jaxesn — CHANGES_REQUESTED** [review]
Left some comments on a couple of the dockerfiles. Some of the comments would apply to the others too.

In general, Im not in love with having to use the standard base because we need bash for all these entrypoint scripts.  If these entrypoints are critical, one thing we may want to look at is building a new minimal base with bash and few scripting tools needed to make these entrypoint script files work as is.

File: projects/goharbor/harbor/Makefile
Comment (jaxesn) [inline projects/goharbor/harbor/Makefile:90]: I ran this locally, looks like it creates files under `src/server/v2.0/models` so can we make this a real target instead? Something like `$(REPO)/src/server/v2.0/models:....`

File: projects/goharbor/harbor/Makefile
Comment (jaxesn) [inline projects/goharbor/harbor/Makefile:1]: To fix gather lionesses/attribution, this project pulls in some json libs from xeipuuv which is not setup exactly how go-licenses expects.  We have handled this in the flux controllers since they pull in the same deps.  If you add something like this to this makefile, it should fix it up.  I verified this locally.

https://github.com/aws/eks-anywhere-build-tooling/blob/main/projects/fluxcd/source-controller/Makefile#L33

File: projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile
Comment (vincentni) [reply projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile:25]: nope

File: projects/goharbor/harbor/docker/linux/harbor-redis/Dockerfile
Comment (TerryHowe) [inline projects/goharbor/harbor/docker/linux/harbor-redis/Dockerfile:1]: Why are we building this redis image when we have one? Same question for nginx and "harbor-db". I'm not sure fi there are any harbor specific changes to those containers

**Review by TerryHowe — CHANGES_REQUESTED** [review]

**Review by jaxesn — CHANGES_REQUESTED** [review]
This is getting pretty close. I think the biggest things from my perspective:

- need the prow jobs so we can see it building and passing
- remove the precompile_prepare
- fix licenses/attribution
- double check the images you are building and see if we can use the existing redis/nginx images, or use the same pattern for those.

File: projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile
Comment (vincentni) [reply projects/goharbor/harbor/docker/linux/harbor-core/Dockerfile:23]: It doesn't work as harbor folder is in root:root otherwise.

File: projects/goharbor/harbor/docker/linux/harbor-portal/Dockerfile
Comment (vincentni) [reply projects/goharbor/harbor/docker/linux/harbor-portal/Dockerfile:1]: oh yeah, I need to update the comment

**Review by jaxesn — APPROVED** [review]
Great work!  This looks pretty good to me.  As a potential item once everything is validated and working, we could circle back and see if we tighten up the base image use here.  Either way, I am good to go on this once we get the prow jobs running.

**Review by TerryHowe — APPROVED** [review]
/lgtm

File: projects/goharbor/harbor/docker/linux/harbor-redis/Dockerfile
Comment (TerryHowe) [inline projects/goharbor/harbor/docker/linux/harbor-redis/Dockerfile:3]: I don't want to hold this up because of this problem, but I want to hear what happened here.

Comment (vincentni) [reply projects/goharbor/harbor/docker/linux/harbor-redis/Dockerfile:3]: I followed the same steps that we built the existing redis image but replaced the configuration file as the existing redis image failed to work in Harbor eco-system.

File: projects/goharbor/harbor/helm/sedfile.template
Comment (jaxesn) [inline projects/goharbor/harbor/helm/sedfile.template:1]: Not for this PR, but we should consider being able to parameterize the GIT_TAG here so its not something we need update when bumping versions. could be easily forgotten.

File: projects/goharbor/harbor/Makefile
Comment (jaxesn) [inline projects/goharbor/harbor/Makefile:59]: Not for this PR, and i know this is a pattern we are following elsewhere, but I think we could probably introduce a helper here to make this cleaner.

**Review by jaxesn — APPROVED** [review]
I am pretty happy with this.  Again, id like to see some future work to tighten up the packages in the images we are building, but that is def not a blocker.

Couple future thinking comments on the helm chart stuff, but otherwise looks good.

File: projects/goharbor/harbor/Makefile
Comment (TerryHowe) [reply projects/goharbor/harbor/Makefile:59]: I have another PR where I use HELM_IMAGE_LIST instead of this. I want to get rid of this thing.

**Review by TerryHowe — APPROVED** [review]
/lgtm

**Pull request merged**
